### PR TITLE
add test for #2479

### PR DESCRIPTION
**Pull request by millken**
#2479

### CODE REVIEW

File: action/protocol/staking/candidate_buckets_indexer_test.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+		{0, 0, 0},
+		{2, 1111111, 1},
+		{21111111, 1111111, 1333333},
+	}
+	for _, v2 := range tests2 {
+		_, _, err = cbi.GetCandidates(v2.height, v2.offset, v2.limit)
+		require.NoError(err)
+	}
+
+	b, h, err := cbi.GetCandidates(2, 0, 1)
```
Comment (dustinxie) [inline action/protocol/staking/candidate_buckets_indexer_test.go:69]: should use tests[0].height or tests[1].height

File: action/protocol/staking/candidate_buckets_indexer_test.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+	require.NoError(err)
+	require.Equal(h, uint64(2))
+
+	var r iotextypes.CandidateListV2
+	require.NoError(proto.Unmarshal(b, &r))
+	require.Equal(len(r.Candidates), 1)
+	require.Equal(r.Candidates[0].Name, "abc")
```
Comment (dustinxie) [inline action/protocol/staking/candidate_buckets_indexer_test.go:76]: should use tests[0].candidate.Name not "abc"

File: action/protocol/staking/candidate_buckets_indexer_test.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	require.Equal(r.Candidates[0].SelfStakeBucketIdx, uint64(123))
```
Comment (dustinxie) [inline action/protocol/staking/candidate_buckets_indexer_test.go:77]: same

File: action/protocol/staking/candidate_buckets_indexer_test.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+		{0, 0, 0},
+		{2, 1111111, 1},
+		{21111111, 1111111, 1333333},
+	}
+	for _, v2 := range tests2 {
+		_, _, err = cbi.GetBuckets(v2.height, v2.offset, v2.limit)
+		require.NoError(err)
+	}
+
+	b, h, err := cbi.GetBuckets(2, 0, 1)
```
Comment (dustinxie) [inline action/protocol/staking/candidate_buckets_indexer_test.go:126]: same as above

File: action/protocol/staking/candidate_buckets_indexer_test.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+		height        uint64
+		offset, limit uint32
+	}{
+		{0, 0, 0},
+		{2, 1111111, 1},
+		{21111111, 1111111, 1333333},
+	}
+	for _, v2 := range tests2 {
+		_, _, err = cbi.GetCandidates(v2.height, v2.offset, v2.limit)
+		require.NoError(err)
```
Comment (dustinxie) [inline action/protocol/staking/candidate_buckets_indexer_test.go:66]: add some case that would fail GetCandidates()?

Comment (millken) [reply action/protocol/staking/candidate_buckets_indexer_test.go:68]: see line 83-99

Comment (dustinxie) [reply action/protocol/staking/candidate_buckets_indexer_test.go:68]: that is not what's being asked, we don't need to test after DB is closed, pls revert 83-99

what i mean is test GetCandidates() possible failure, for example:
1. when height does not exist, it returns an empty list, then you should test such
2. when offset > length, it returns an empty list, then you should test such

not only require.NoError(err)

File: action/protocol/staking/candidate_buckets_indexer_test.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	b, h, err := cbi.GetCandidates(tests[1].height, 0, 1)
+	require.NoError(err)
+	require.Equal(h, tests[1].height)
```
Comment (dustinxie) [inline action/protocol/staking/candidate_buckets_indexer_test.go:73]: can be combined into the above for() loop

File: action/protocol/staking/candidate_buckets_indexer_test.go [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+		{21111111, 1111111, 1333333},
+	}
+	for _, v2 := range tests2 {
+		_, _, err = cbi.GetBuckets(v2.height, v2.offset, v2.limit)
+		require.NoError(err)
+	}
+
+	b, h, err := cbi.GetBuckets(tests[1].height, 0, 1)
+	require.NoError(err)
+	require.Equal(h, uint64(2))
```
Comment (dustinxie) [inline action/protocol/staking/candidate_buckets_indexer_test.go:148]: is it possible to combine into above for() loop?

File: action/protocol/staking/candidate_buckets_indexer_test.go [hunk 8]
```diff
@@ -54,49 +52,35 @@ func TestCandidatesBucketsIndexer_PutGetCandidates(t *testing.T) {
...
-	require.NoError(proto.Unmarshal(b, &r))
-	require.Equal(len(r.Candidates), 1)
-	require.Equal(r.Candidates[0].Name, tests[1].candidates.Candidates[0].Name)
-	require.Equal(r.Candidates[0].SelfStakeBucketIdx, tests[1].candidates.Candidates[0].SelfStakeBucketIdx)
+		var r iotextypes.CandidateListV2
+		require.NoError(proto.Unmarshal(a, &r))
+		require.Equal(len(r.Candidates), v2.r1)
+		require.Equal(b, v2.r2)
 
+		if len(r.Candidates) > 0 {
```
Comment (dustinxie) [inline action/protocol/staking/candidate_buckets_indexer_test.go:78]: can you do the same for PutGetBuckets(t *testing.T)?

File: action/protocol/staking/candidate_buckets_indexer_test.go [hunk 9]
```diff
@@ -54,49 +52,35 @@ func TestCandidatesBucketsIndexer_PutGetCandidates(t *testing.T) {
...
-	require.NoError(err)
-	require.Equal(h, tests[1].height)
-
-	var r iotextypes.CandidateListV2
-	require.NoError(proto.Unmarshal(b, &r))
-	require.Equal(len(r.Candidates), 1)
-	require.Equal(r.Candidates[0].Name, tests[1].candidates.Candidates[0].Name)
-	require.Equal(r.Candidates[0].SelfStakeBucketIdx, tests[1].candidates.Candidates[0].SelfStakeBucketIdx)
+		var r iotextypes.CandidateListV2
+		require.NoError(proto.Unmarshal(a, &r))
```
Comment (dustinxie) [inline action/protocol/staking/candidate_buckets_indexer_test.go:74]: can you do the same for PutGetBuckets(t *testing.T)?

including the if() check below

File: action/protocol/staking/candidate_buckets_indexer_test.go [hunk 10]
```diff
@@ -54,49 +52,35 @@ func TestCandidatesBucketsIndexer_PutGetCandidates(t *testing.T) {
...
+		{2, 0, 1, 1, 2},
+		{2, 0, 100, 1, 2},
+		{2, 1234, 0, 0, 2},
+		{2, 1111111, 1, 0, 2},
+		{21111111, 1111111, 1333333, 0, 2},
 	}
 	for _, v2 := range tests2 {
-		_, _, err = cbi.GetCandidates(v2.height, v2.offset, v2.limit)
+		a, b, err := cbi.GetCandidates(v2.height, v2.offset, v2.limit)
+		t.Logf("%v, a=%s, b= %d", v2, a, b)
```
Comment (dustinxie) [inline action/protocol/staking/candidate_buckets_indexer_test.go:70]: remove log

**Review by dustinxie — APPROVED** [review]

**Pull request merged**